tiledb/cf/netcdf_engine/_array_converters.py

```diff
@@ -124,62 +187,45 @@ def copy(
                 if isinstance(attr_creator, NetCDF4ToAttrConverter):
                     attr_creator.copy_metadata(netcdf_group, tiledb_array)
             for dim_creator in self._domain_creator:
-                if isinstance(dim_creator.base, NetCDF4ToDimConverter):
+                if isinstance(dim_creator.base, NetCDF4ToDimBase):
                     dim_creator.base.copy_metadata(netcdf_group, tiledb_array)
-            # Copy array data to TileDB.
-            if self.sparse:
-                shape: Optional[Union[int, Sequence[int]]] = -1
-            else:
-                shape = (
-                    None
-                    if all(
-                        isinstance(dim_creator.base, NetCDF4ToDimConverter)
-                        for dim_creator in self._domain_creator
-                    )
-                    else self.domain_creator.get_dense_query_shape(netcdf_group)
-                )
-            data = {}
-            for attr_creator in self:
-                if isinstance(attr_creator, NetCDF4ToAttrConverter):
-                    data[attr_creator.name] = attr_creator.get_values(
-                        netcdf_group, sparse=self.sparse, shape=shape
-                    )
-                else:
-                    if (
-                        assigned_attr_values is None
-                        or attr_creator.name not in assigned_attr_values
-                    ):
-                        raise KeyError(
-                            f"Missing value for attribute '{attr_creator.name}'."
-                        )
-                    data[attr_creator.name] = assigned_attr_values[attr_creator.name]
-            coord_values = self._domain_creator.get_query_coordinates(
-                netcdf_group, self.sparse, assigned_dim_values
+            # Copy array data for first fragment.
+            self._copy_to_array(
+                netcdf_group,
+                tiledb_array,
+                next(fragment_indexers),
+                assigned_dim_values,
+                assigned_attr_values,
             )
-            tiledb_array[coord_values] = data
+        # Copy array data for remaining fragments.
+        for indexer in fragment_indexers:
+            with tiledb.open(
+                tiledb_uri, mode="w", key=tiledb_key, ctx=tiledb_ctx
+            ) as tiledb_array:
```

Comment:
- Why does the first fragment have to be copied separately?
- Why does it need to open (and close) the array on every iteration?

Reply:
Opening (and closing) the array multiple times is what creates multiple discrete fragments. The first step is separate since the metadata is only copied once.

Reply:
Unless something changed recently, I believe a fragment is created on every write to the array, no need to reopen/close it; @ihnorton is this right?

Reply:
You are right. I'll update that.